### ISSUE TITLE
Why is mrcnn_class_loss masked by active_class_ids?

### ISSUE THREAD

**Issue by spotiris** [matterport/Mask_RCNN#1467]
In the current MRCNN implementation, the mrcnn_class_loss is only calculated for the "active_class_ids" of the "source.id" mapping (from the utils.Dataset parent class) for each image (really each batch as there is a TODO in the model).

Does this not mean if you have multiple sources the classifier head will not learn from negative examples of other classes?
If so, is there any reason to keep this mechanism there?

I have tested this by modifying MRCNN to enable all classes in every source.
I trained the standard MRCNN and my modification with the same hyperparameters with my dataset that has two sources. I used ReduceLROnPlateau(patience=10, cooldown=20) starting at 2e-3.

The results was that every training loss was significantly larger when using class_id masking by source.
These results were also reflected in the validation losses.

Below is an image of the training loss. The light blue line is using the standard MRCNN (with class_id source masking) and the dark blue line is without.

![image](https://user-images.githubusercontent.com/13250749/56707855-f2860680-675d-11e9-91d2-2e012e7730a2.png)

**Comment by bertsky** [comment]
@spotiris first of all, I would not call that difference significant :-) (especially since you don't show us the validation loss, too)

Also, your observation that the `active_class_ids` mechanism would exclude any classes from the loss that are not in the _current batch_ is not correct: It's always activating all classes of the whole `dataset` passed to the training generator.

The reason for having this mechanism in place is that you might know in advance that some classes don't appear (at all) in a dataset, while they do in others. Then it is beneficial for the discriminative capacity of the model to not allow the gradient of the impossible classes to contribute to the loss – because this would "spoil" these classes over the course of this dataset. Thus, it's not necessary to show the network instances of all classes at all times. (Remember the classifier head cannot "learn" from negatives only, it needs some balance between positives and negatives.)

But properly learning background (as a class) is also necessary (and hard if images/crops are sparse or empty because this implementation feeds as few negatives as there are positives so BG can only be learned indirectly via zero padding – see this [here](https://github.com/matterport/Mask_RCNN/pull/1088#issuecomment-695208655) for a solution.)

**Comment by bertsky** [comment]
(The mechanism does not work in the `master` version BTW, because it uses `dataset.num_classes` instead of `config.NUM_CLASSES` required to get the correct network shape for the `image_meta` input.)
